### ISSUE TITLE
Rule S2743: FP for static fields used in lock statements

### ISSUE THREAD

**Issue by costin-zaharia-sonarsource** [SonarSource/sonar-dotnet#3839]
### Description

Source: https://community.sonarsource.com/t/csharpsquid-s2743-static-fields-should-not-be-used-in-generic-types-on-lock-object/33552

### Repro steps
```
using System.Collections.Generic;

namespace Stryker.Core.Initialisation
{
    public class FolderCompositeCache<T>
    {
        private FolderCompositeCache()
        {

        }

        private static FolderCompositeCache<T> _instance;
        private static readonly object _lockObj = new object();

        public static FolderCompositeCache<T> Instance
        {
            get
            {
                lock (_lockObj)
                {
                    if (_instance == null)
                    {
                        _instance = new FolderCompositeCache<T>();
                    }

                    return _instance;
                }
            }
        }

        public Dictionary<string, T> Cache { get; set; }
    }
}
```

### Expected behavior

Rule should not trigger in this case.

### Actual behavior

S2743 is raising an issue for `_lockObj`

### Known workarounds

Use `new T()` but this is not always possible not recommended. 

### Related information

* C#/VB.NET Plugins Version: 8.15

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1933](https://sonarsource.atlassian.net/browse/NET-1933)

[NET-1933]: https://sonarsource.atlassian.net/browse/NET-1933?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
